Add NextFTC Pedro Extension

BeepBot99 opened the pull request:
Adds `PedroComponent` and the following commands:

- `FollowPath`
- `PedroDriverControlled`
- `Turn`
- `TurnTo`

Proximity and displacement delays were not added because:

1. Those can be implemented using `WaitUntil`.
2. The goal of this extension is just to mirror the features of Pedro Pathing using NextFTC, not to add any more.

BeepBot99 committed:
refactor: remove unit and instrumented tests from pedro module

BeepBot99 committed:
chore: change group id to dev.nextftc.extensions

BeepBot99 committed:
feat: add PedroComponent

BeepBot99 committed:
chore: remove unused import in build.gradle.kts

BeepBot99 committed:
feat: add FollowPath

BeepBot99 committed:
feat: add PedroComponent.gyro

BeepBot99 committed:
feat: add PedroDriverControlled

BeepBot99 committed:
feat: add Turn and TurnTo

zachwaffle4 reviewed (changes requested):
Everything should be in `src/main/kotlin/dev/nextftc/extensions/pedro`, not `src/main/dev/nextftc/extensions/pedro`.

BeepBot99 committed:
fix: move files from src/main to src/main/kotlin

BeepBot99 commented:
> Everything should be in `src/main/kotlin/dev/nextftc/extensions/pedro`, not `src/main/dev/nextftc/extensions/pedro`.

Fixed, thank you.

zachwaffle4 commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/PedroComponent.kt:
```diff
@@ -0,0 +1,36 @@
+    }
+}
+
+private var _follower: Follower? = null
```
I would rather users just pass their `Follower` into the component; this implementation is very strange (having a class whose sole purpose is to manage a top-level property).

BeepBot99 replied:
Would you prefer it be in a companion object?

zachwaffle4 commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/Turn.kt:
```diff
@@ -0,0 +1,15 @@
+import dev.nextftc.core.units.Angle
+import dev.nextftc.core.units.abs
+
+class Turn(private val angle: Angle) : Command() {
```
It might be useful to name this like `"Turn($angle)"`.
```suggestion
class Turn(private val angle: Angle) : Command() {
  init {
    named("Turn($angle)"
  }
```

BeepBot99 replied:
Right, I forgot names existed. This suggestion is missing a closing parenthesis, but I can add this tomorrow.

zachwaffle4 commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/TurnTo.kt:
```diff
@@ -0,0 +1,14 @@
+import dev.nextftc.core.commands.Command
+import dev.nextftc.core.units.Angle
+
+class TurnTo(private val angle: Angle) : Command() {
```
Similarly it might be wise to name this `"TurnTo($angle)"`.

zachwaffle4 commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/PedroComponent.kt:
```diff
@@ -0,0 +1,36 @@
+import dev.nextftc.ftc.ActiveOpMode
+import java.util.function.Supplier
+
+class PedroComponent(private val followerFactory: (hardwareMap: HardwareMap) -> Follower) : Component {
```
I think this should be an `object` that is added to an `OpMode` using `PedroComponent.withFollower(...)`;

```suggestion
object PedroComponent : Component {
  lateinit var follower: Follower
    private set

  fun withFollower(follower: Follower) = apply {
    this.follower = follower
  }
```

BeepBot99 replied:
Why do you think it should be a singleton?

zachwaffle4 replied:
If `follower` is a globally accessible property (and not a member property), it should be managed by a globally accessible object and not a class.

BeepBot99 replied:
Okay.

BeepBot99 committed:
feat: add descriptive name to Turn and TurnTo

- name of Turn is Turn(angle)
- name of TurnTo is TurnTo(angle)

BeepBot99 committed:
refactor: change PedroComponent to an object

zachwaffle4 commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/PedroComponent.kt:
```diff
@@ -0,0 +1,38 @@
+@file:JvmName("FollowerHolder")
```
is this necessary now that the file itself no longer has any top-level properties?

rowan-mcalpin replied:
It's not needed; @BeepBot99 please remove

BeepBot99 replied:
No, I'm going to add it back again.

rowan-mcalpin commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/FollowPath.kt:
```diff
@@ -0,0 +1,37 @@
+) : Command() {
+
+    init {
+        require(maxPower == null || holdEnd != null) { "If maxPower is passed, holdEnd must be passed as well." }
```
Is there a reason not to default holdEnd to true?

BeepBot99 replied:
Yes, because the default `holdEnd` value is configurable in `FollowerConstants`. It should use that value when not passed, but there is no overload of `followPath` that takes `maxPower` but not `holdEnd`.

rowan-mcalpin commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/PedroDriverControlled.kt:
```diff
@@ -0,0 +1,26 @@
+    drivePower: Supplier<Double>,
+    strafePower: Supplier<Double>,
+    turnPower: Supplier<Double>,
+    private val robotCentric: Boolean = true
```
IMO, should default to false

BeepBot99 replied:
Robot centric is a more sensible default and is what Pedro defaults to.

rowan-mcalpin commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/Turn.kt:
```diff
@@ -0,0 +1,20 @@
+import dev.nextftc.core.units.abs
+import dev.nextftc.extensions.pedro.PedroComponent.follower
+
+class Turn(private val angle: Angle) : Command() {
```
Can you rename this to `TurnBy` to better contrast with `TurnTo`?

BeepBot99 replied:
In Pedro, it's called `turn`. I'd like to stay consistent with that.

rowan-mcalpin replied:
While consistency is valuable, I'm going to prioritize clarity in our library.

zachwaffle4 commented on pedro/build.gradle.kts:
```diff
@@ -30,9 +30,6 @@ dependencies {
-    testImplementation(libs.mockk)
 }
 
 description =
```
please fix the publishing stuff to *not* be the Hardware description + name

zachwaffle4 reviewed (changes requested):
Please add the `versions.pedro` property.

BeepBot99 committed:
fix: make PedroComponent a class again

BeepBot99 committed:
fix(pedro): fix gradle

- make displayName and description correct
- use versions.pedro property for version

zachwaffle4 commented on gradle.properties:
```diff
@@ -2,5 +2,5 @@ kotlin.code.style=official
 org.jetbrains.dokka.experimental.gradle.pluginMode.noWarn=true
 
-version=1.0.0
+versions.pedro=1.0.0
```
this should be 2.0.0-1 according to the current extension version scheme

rowan-mcalpin reviewed (changes requested):
Rename `Turn` to `TurnBy`

BeepBot99 committed:
refactor: rename Turn to TurnBy

rowan-mcalpin commented on pedro/src/main/kotlin/dev/nextftc/extensions/pedro/TurnBy.kt:
```diff
@@ -0,0 +1,20 @@
+class TurnBy(private val angle: Angle) : Command() {
+
+    init {
+        named("Turn($angle)")
```
Can you update the name as well? (To `TurnBy($angle)`)

BeepBot99 committed:
fix: rename Turn to TurnBy in the command name

rowan-mcalpin reviewed (approved):
Thanks!

zachwaffle4 committed:
Merge remote-tracking branch 'origin/develop' into feature/pedro-extension

# Conflicts:
#	build.gradle.kts
#	gradle.properties
#	pedro/src/test/java/dev/nextftc/extensions/ExampleUnitTest.kt
